### PR TITLE
Feature/registration_authentication

### PR DESCRIPTION
**Pull request by TeodoraRoksana**
Added Registration and Authentication. 
Added PasswordSalt table which imitates a different storage for higher security.
Added UserController.
Added Services Project (DBService, HashService).
Added password hashing. 

Changed User table: added Username field; changed saving of Age to DateOfBirth; changed type of password hash (was byte[] -> now string).

### CODE REVIEW

**Review by Brain-DedW — APPROVED** [review]
No problems

**Review by sasha-sush — APPROVED** [review]
Everything's ok

File: TWC_Backend/TWC_Backend/Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                if (!ModelState.IsValid)
+                    throw new Exception("Invalid data. Please check the data. Email must be in the correct format and password must be longer than 8 characters!");
+
+                var user = await _dBService.GetUserByEmail(userAuthenticationDTO.Email);
+
+                if (user == null)
+                    throw new Exception("User with this email does not exist!");
+
+                var salt = await _dBService.GetSaltByUserId(user.Id);
+                //bool isPasswordCorect =
```
Comment (koliahreb) [inline TWC_Backend/TWC_Backend/Controllers/UserController.cs]: remove

File: TWC_Backend/TWC_DatabaseLayer/Models/Chat.cs [hunk 2]
```diff
@@ -12,8 +12,8 @@ public class Chat
         [Key]
         public int Id { get; set; }
         public int ProjectId { get; set; }
-        public int UserId1 { get; set; }
-        public int UserId2 { get; set; }       
+        public int User1Id { get; set; }
+        public int User2Id { get; set; }
```
Comment (koliahreb) [inline TWC_Backend/TWC_DatabaseLayer/Models/Chat.cs:11]: try to use many to many in next pr

File: TWC_Backend/TWC_DatabaseLayer/Models/PasswordSalt.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.ComponentModel.DataAnnotations;
```
Comment (koliahreb) [inline TWC_Backend/TWC_DatabaseLayer/Models/PasswordSalt.cs]: some of them not used

File: TWC_Backend/TWC_Services/DBService/DBService.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+{
+    public class DBService : IDBService
+    {
+        private DataContext _context;
+
+        public DBService(DataContext context) {  _context = context; }
+
+        public async Task<List<User>> GetAllUsers()
+        {
+            return (await _context.Users.ToListAsync());
```
Comment (koliahreb) [inline TWC_Backend/TWC_Services/DBService/DBService.cs]: remove brackets

File: TWC_Backend/TWC_Services/DBService/DBService.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            {
+                throw new Exception($"cant found {typeof(User).Name} by email. Messege error: " + ex.Message);
+            }
+        }
+
+        public async Task<User> AddUser(User user)
+        {
+            try
+            {
+                //var checkUser = await GetUserByEmail(user.Email);
```
Comment (koliahreb) [inline TWC_Backend/TWC_Services/DBService/DBService.cs]: remove

File: TWC_Backend/TWC_Services/DBService/DBService.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                _context.Users.Remove(user);
+                await _context.SaveChangesAsync();
+            }
+            catch (Exception ex)
+            {
+                throw new Exception($"cant delete {typeof(User).Name}. Messege error: " + ex.Message);
+            }
+        }
+
+        public async Task<PasswordSalt> AddSalt(PasswordSalt salt)
```
Comment (koliahreb) [inline TWC_Backend/TWC_Services/DBService/DBService.cs]: Async suffix for async methods

**Review by koliahreb — CHANGES_REQUESTED** [review]

**Review by koliahreb — APPROVED** [review]
looks good

File: TWC_Backend/TWC_Services/DBService/DBService.cs (hunk 6 above)
Comment (koliahreb) [reply TWC_Backend/TWC_Services/DBService/DBService.cs]: done - this comment should be from person who added code to resolve this comment)

File: TWC_Backend/TWC_Backend/Controllers/UserController.cs (hunk 1 above)
Comment (VladyslavZhyliakov) [reply TWC_Backend/TWC_Backend/Controllers/UserController.cs]: removed

File: TWC_Backend/TWC_Services/DBService/DBService.cs (hunk 6 above)
Comment (VladyslavZhyliakov) [reply TWC_Backend/TWC_Services/DBService/DBService.cs]: Added suffixes for all async methods

**Comment by StepanZdebskyi** [comment]
Well done!

**Pull request merged**